doc/6.externs/taskobj.c

```diff
@@ -0,0 +1,390 @@
+                #endif
+                    remaining -= delta;
+                }
+            }
+            else /* fail */
+            {
+                x->err = errno;
+                free(data);
+            }
+        }
+    }
+    else /* catch error */
+        x->err = errno;
+}
+
+    /* this function is called when the task
+     * has been completed or cancelled */
+static void taskobj_read_callback(t_taskobj *x, t_read_data *y)
+{
+    if (x) /* completed */
+    {
+            /* move file data to object */
+        x->x_data = y->data;
+        x->x_size = y->size;
+            /* the task handle is now invalid, so we have to unset it!
+             * Let's do this BEFORE sending the message, so the user
+             * can call the "read" method recursively. */
+        x->x_task = 0;
+            /* notify */
+        if (x->x_data) /* success */
+            outlet_float(x->x_statusout, x->x_size);
+        else /* fail */
+        {
+            pd_error(x, "could not read file '%s': %s (%d)",
+                y->filepath, strerror(y->err), y->err);
+            outlet_float(x->x_statusout, -1);
+        }
+    }
+    else /* cancelled */
+    {
+        post("taskobj: task has been cancelled");
+            /* free any file data to avoid a memory leak!
+             * NB: we could use task_sched() to defer resource cleanup
+             * to the worker thread, but this is omitted for brevity. */
+        if (y->data)
+            free(y->data);
+    }
+        /* finally free the task data object */
+    freebytes(y, sizeof(t_read_data));
+}
+
+static void taskobj_read(t_taskobj *x, t_symbol *file, t_floatarg ms)
+{
+    t_read_data *y;
+        /* check if a task is still running; typically it does not
+         * make sense to run several tasks concurrently */
+    if (x->x_task)
+    {
+    #if TASKOBJ_CANCEL
+        taskobj_cancel(x);
+    #else
+            /* treat as error and return */
+        pd_error(x, "taskobj: task still running");
+        return;
+    #endif
+    }
+        /* allocate task data */
+    y = (t_read_data *)getbytes(sizeof(t_read_data));
+    y->err = 0;
+    y->ms = ms;
+        /* move old file data, to be deleted in the worker thread */
+    y->data = x->x_data;
+    y->size = x->x_size;
+    x->x_data = 0;
+    x->x_size = 0;
+        /* store file name */
+    snprintf(y->filepath, sizeof(y->filepath), "%s", file->s_name);
+        /* schedule task and store task handle */
+    x->x_task = task_sched((t_pd *)x, y, (t_task_workfn)taskobj_read_workfn,
+        (t_task_callback)taskobj_read_callback);
+}
+
+/* ------------------- taskobj_download --------------------- */
+
+typedef struct _download_data
+{
+    char url[1024];
+    char filepath[256];
+    int success;
+} t_download_data;
+
+static void taskobj_download_workfn(t_task *task, t_download_data *x)
+{
+    char cmd[MAXPDSTRING];
+    snprintf(cmd, sizeof(cmd), "curl %s --output %s", x->url, x->filepath);
```

Comment:
Thanks. I reduced the size of the `url` field instead. "Real" code would need to heap-allocate the URL and file path (instead of imposing arbitrary size limits), but I wanted to keep it as simple as possible...

Reply:
a "real" object will hopefully use `execv` instead of `system`...

Reply:
@umlaeute I think this is addressed by the code comment :-)
```
/* we're lazy and just call the 'curl' command with 'system()'.
 * In reality, you should probably use libcurl instead :-) */
 ```

Reply:
Fair enough 😉